drivers: Add BMM150 geomagnetic sensor driver

punitvara opened the pull request:
This patch implements BMM150 driver with polling mode along with sample
application to test it. Driver works on I2C interface as of now.

Following datasheet is taken as reference while developing driver.
http://www.mouser.com/ds/2/783/BST-BMM150-DS001-01-786480.pdf

Signed-off-by: Punit Vara<punit.vara@intel.com>

nashif commented:
> JIRA-ZEP-1939

Jira: ZEP-1939

> Signed-off-by: Punit Vara<punit.vara@intel.com>

add space. Are you adding this manually? This should be automatically added when you commit using -s.


Please fix all those checpatch errors and make sure you add the hook, look at https://github.com/zephyrproject-rtos/zephyr/wiki/Contribution-Guide#coding-style

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+		rep_z = data->rep_z;
+	}
+
+	*max_odr = 1000000 / (145 * rep_xy + 500 * rep_z + 980);
```
This equation needs some explanation. What are all the magic numbers?

punitvara replied:
That's equation for maximum selectable read-out frequency in forced mode.  Its given in datasheet. I will mention datasheet reference for reference.

MaureenHelm replied:
Please add a comment in the code that says this equation comes directly from the datasheet

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+              ((((s32_t)val) * ((s32_t)val)) >> 7)) + (((s32_t)val) *
+              ((s32_t)(((s16_t)tregs->xy1) << 7)))) >> 9) +
+              ((s32_t)0x100000)) * ((s32_t)(((s16_t)txy2) +
+              ((s16_t)0xA0)))) >> 12)) >> 13)) + (((s16_t)txy1) << 3);
```
These equations are really difficult to read. Can you please break down into smaller pieces and clean up the magic numbers?

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+	val = (((((s32_t)(z - tregs->z4)) << 15) - ((((s32_t)tregs->z3) *
+              ((s32_t)(((s16_t)rhall) - ((s16_t)tregs->xyz1)))) >> 2)) /
+              (tregs->z2 + ((s16_t)(((((s32_t)tregs->z1) *
+              ((((s16_t)rhall) << 1))) + (1 << 15)) >> 16))));	
```
This equation is really difficult to read. Can you please break down into smaller pieces and clean up the magic numbers?

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+						  true);
+	drv_data->sample_y = bmm150_compensate_xy(&drv_data->tregs, raw_y, rhall,
+						  false);
+ 	drv_data->sample_z = bmm150_compensate_z(&drv_data->tregs, raw_z, rhall);
```
Why is compensate separate from convert? I think it should all be done together in the channel_get function.

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+#if defined(CONFIG_BMM150_SAMPLING_REP_XY)
+        case SENSOR_CHAN_MAGN_X:
+        case SENSOR_CHAN_MAGN_Y:
+                if (val->val1 < 1 || val->val1 > 511) {
```
What do the magic numbers mean? Can you replace them with descriptive macros?

punitvara replied:
MAX repetitions value for XY and Z axis. Converted to Macros. Done

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+        }
+
+        if (i2c_reg_read_byte(data->i2c, config->i2c_slave_addr,
+                              BMM150_REG_CHIP_ID, &chip_id) < 0) {
```
Shouldn't this be the very first thing you do, before setting power modes?

punitvara replied:
According to Datasheet section 5.3 
This number can only be read if the power control bit (register 0x4B bit0) is enabled.

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+                return -EIO;
+        }
+
+        dev->driver_api = &bmm150_api_funcs;
```
You can use DEVICE_AND_API_INIT instead of DEVICE_INIT below, then remove this line

punitvara replied:
Done

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+
+static const struct bmm150_config bmm150_config = {
+        .i2c_master_dev_name = CONFIG_BMM150_I2C_MASTER_DEV_NAME,
+        .i2c_slave_addr = BMM150_I2C_ADDR,
```
You can use CONFIG_BMM150_I2C_ADDR directly and remove the extra BMM150_I2C_ADDR macro

MaureenHelm commented on drivers/sensor/bmm150/bmm150.h:
```diff
@@ -0,0 +1,200 @@
+ * * distributed under the License is distributed on an "AS IS" BASIS,
+ * * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * * See the License for the specific language governing permissions and
+ * * limitations under the License.
```
See my earlier comment about SPDX

punitvara replied:
Done

MaureenHelm commented on drivers/sensor/bmm150/bmm150.h:
```diff
@@ -0,0 +1,200 @@
+	char *i2c_master_dev_name;
+	u16_t i2c_slave_addr;
+
+#if defined(CONFIG_BMM150_TRIGGER_DRDY)
```
It doesn't appear that you've implemented any trigger support yet. Can you remove this and all other trigger references?

punitvara replied:
Done

MaureenHelm replied:
This still needs to be removed.

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+{
+	/* val = raw_val /1600 */
+	val->val1 = raw_val / 1600;
+	val->val2 = ((s32_t)raw_val * (1000000 / 1600)) % 1000000;
```
Please add some explanation to this equation. What are the raw_val units? What are the output units (they should be Gauss)?

punitvara replied:
We fetch raw values in samples_fetch API. Do I still need to comment and describe it ?

punitvara replied:
I have taken this reference from another driver bmc_magc

MaureenHelm replied:
It helps to understand the conversion equation if you explain what units you're converting from and to. It's sufficient to say that the datasheet specifies raw units are 16 LSB/uT and you're converting to Gauss.

MaureenHelm commented on samples/sensor/bmm150/src/main.c:
```diff
@@ -0,0 +1,79 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Please move the sample to a separate commit

punitvara replied:
Will split commit once above comments are fixed

dbkinder commented on samples/sensor/bmm150/README.rst:
```diff
@@ -0,0 +1,8 @@
+BMM150 GeoMagnetic Sensor
```
Please use the example template for sample code found in your cloned zephyr tree: doc/templates/sample.tmpl

punitvara replied:
Done

dbkinder replied:
Hmmm... I only see a title and intro sentence.  Is more coming?

dbkinder reviewed (changes requested):
Please use the example template for sample code found in your cloned zephyr tree: doc/templates/sample.tmpl

punitvara commented on drivers/sensor/bmm150/Kconfig:
```diff
@@ -0,0 +1,96 @@
+config BMM150_I2C_MASTER_DEV_NAME
+	string "I2C master where BMM150_MAGN is connected"
+	depends on BMM150
+	default "I2C_0"
```
What should be "NAME" here ? Most of the sensor drivers Kconfig contains I2C_0 . Should I use any arbitrary NAME there ?

punitvara replied:
@MaureenHelm  Can you please reply and review further driver code so that I can merge for 1.9 ?

punitvara replied:
Ok. So not changing. thanks for pointing that

punitvara commented on drivers/sensor/bmm150/Kconfig:
```diff
@@ -0,0 +1,96 @@
+	help
+	  Enable alteration of sampling rate attribute at runtime.
+
+config BMM150_SAMPLING_REP_XY
```
REP is for number of representations

punitvara commented on drivers/sensor/bmm150/Kconfig:
```diff
@@ -0,0 +1,96 @@
+	  Specify the I2C slave address for the BMM150
+
+config BMM150_I2C_MASTER_DEV_NAME
+	string "I2C master where BMM150_MAGN is connected"
```
Done

punitvara replied:
Pushed again. Done

punitvara commented on drivers/sensor/bmm150/Kconfig:
```diff
@@ -0,0 +1,96 @@
+
+config BMM150_PRESET_LOW_POWER
+	bool
+	prompt "Low power (3, 3, 10)"
```
Done

punitvara commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+						!state) < 0) {
+				return -EIO;
+			}
+			k_busy_wait(5 * USEC_PER_MSEC);
```
OK. Removed this delay as of no. I took the reference from another zephyr driver but not needed now

punitvara replied:
Without this I am not able get proper output. I took reference from another similar driver bmc** in zephyr.

punitvara commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+						   BMM150_REG_OPMODE_ODR,
+						   BMM150_MASK_OPMODE,
+						   BMM150_MODE_SLEEP <<
+						   BMM150_SHIFT_OPMODE);
```
Done

punitvara commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+	u8_t i;
+
+	for(i=0; i < ARRAY_SIZE(bmm150_samp_freq_table); ++i) {
+		if (val <= bmm150_samp_freq_table[i].freq) {
```
Done

punitvara commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+						   BMM150_MASK_ODR,
+						   (bmm150_samp_freq_table[i].
+						   reg_val <<
+						   BMM150_SHIFT_ODR));
```
Done

punitvara commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+		return -EIO;
+	}
+
+	data->rep_xy = BMM150_REGVAL_TO_REPXY((int)(reg_val));
```
Done

tbursztyka commented on drivers/sensor/bmm150/Kconfig:
```diff
@@ -0,0 +1,81 @@
+	default n
+	help
+	  Enable driver for BMM150 I2C-based Geomagnetic sensor.
+
```
add: 
if BMM150

punitvara replied:
Done

tbursztyka commented on drivers/sensor/bmm150/Kconfig:
```diff
@@ -0,0 +1,81 @@
+	help
+	  Enable alteration of Z oversampling at runtime.
+
+endmenu
```
- remove  endmenu
- add endif # BMM150

punitvara replied:
Done

tbursztyka commented on drivers/sensor/bmm150/Kconfig:
```diff
@@ -0,0 +1,81 @@
+	  Specify the device name of the I2C master device to which BMM150
+	  is connected.
+
+menu "Attributes"
```
remove this (and the depends below)

punitvara replied:
Done

tbursztyka commented on drivers/sensor/bmm150/Kconfig:
```diff
@@ -0,0 +1,81 @@
+
+config BMM150_DEV_NAME
+	string "BMM150 device name"
+	depends on BMM150
```
remove all the "depends on BMM150" in this file, the if BMM150 above will replace those.

punitvara replied:
Done

tbursztyka commented on drivers/sensor/bmm150/bmm150.h:
```diff
@@ -0,0 +1,167 @@
+#include <gpio.h>
+
+
+#define BMM150_REG_CHIP_ID         0x40
```
Use tabs to align defined names and value, do that everywhere

punitvara replied:
Done

tbursztyka commented on drivers/sensor/bmm150/bmm150.h:
```diff
@@ -0,0 +1,167 @@
+#define BMM150_XY_OVERFLOW_VAL     -4096
+#define BMM150_Z_OVERFLOW_VAL      -16384
+
+#define BMM150_REP_XY_MIN	   1
```
for instance here you would not have had this tiny alignment issue with tabs

tbursztyka commented on drivers/sensor/bmm150/bmm150.h:
```diff
@@ -0,0 +1,167 @@
+	struct device *i2c;
+	struct k_sem sem;
+	struct bmm150_trim_regs tregs;
+	int rep_xy, rep_z, odr, max_odr;
```
explode this
int rep_xy;
int rep_z;
... etc

do the same below for sample_*

tbursztyka commented on drivers/sensor/bmm150/bmm150.h:
```diff
@@ -0,0 +1,167 @@
+};
+
+enum bmm150_presets {
+	LOW_POWER_PRESET,
```
Add BMM150_ prefix

punitvara replied:
Done

tbursztyka commented on samples/sensor/bmm150/src/main.c:
```diff
@@ -0,0 +1,79 @@
+		ret = sensor_channel_get(dev, SENSOR_CHAN_MAGN_Z, &z);
+		
+		printf("( x y z ) = ( %f  %f  %f )\n",
+                       sensor_value_to_double(&x),
```
one space too many for the alignment (unless it's github's view which makes it weird). These should be aligned right below the "

tbursztyka commented on samples/sensor/bmm150/src/main.c:
```diff
@@ -0,0 +1,79 @@
+			printk("device binding \n");
+			return dev;
+		}
+		++i;
```
break a line

tbursztyka commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,602 @@
+static const struct {
+	int freq;
+	u8_t reg_val;
+} bmm150_samp_freq_table[] = { {2, 0x01},
```
bad alignment:
`} bmm150_samp_freq_table[] = {
        { 2, 0x01 },
        { 6, 0x02 },
...
};
`

Of course with one tab instead of the 8 space I had to use here.

punitvara replied:
Done

tbursztyka commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,602 @@
+	u8_t rep_z;
+	u8_t odr;
+} bmm150_presets_table[] = {
+	[LOW_POWER_PRESET] = {3 , 3, 10},
```
space after { and before }

tbursztyka commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,602 @@
+						   BMM150_SHIFT_OPMODE) < 0) {
+				return -EIO;
+			}
+			return 0;
```
Break a line, I am picky about that but lines breaks are cheap and is makes the code breathing a bit, more readable imo. (rule being: break a line after a '}' unless it's another '}' or an else etc..). Apply that everywhere.

tbursztyka commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,602 @@
+};
+
+static int bmm150_set_power_mode(struct device *dev,
+				enum bmm150_power_modes mode,
```
parameters alignment

foo(param1, ...
      paramN, ..., paramK);

That goes for function declaration as well as function calls. Please apply that everywhere (your editor should be able to handle that automatically btw)

tbursztyka commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,602 @@
+	}
+	
+	return -ENOTSUP; 
+
```
remove that empty line.

tbursztyka commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,602 @@
+						   BMM150_MASK_OPMODE,
+						   BMM150_MODE_NORMAL <<
+						   BMM150_SHIFT_OPMODE) < 0) {
+				return -EIO;
```
you could return i2c call directly, and thus removing return -EIO  and return 0.
Here it makes sense to do so, as you are not doing anything after this call.
Apply that in other relevant places as well.

tbursztyka commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,602 @@
+		rep_z = data->rep_z;
+	}
+	
+	/* Datasheet 4.2.4 Equation Refernce */
```
typo: Reference

tbursztyka commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,602 @@
+	case SENSOR_CHAN_MAGN_X:
+		drv_data->sample_x = bmm150_compensate_xy(&drv_data->tregs, 
+				drv_data->sample_x,drv_data->rhall,
+                                                  true);
```
here, as you cannot align the parameters to last opened '(' (due to 80 chars limit), you still need to align subsequent parameters to each other: so 'true' should be below drv_data->sample_...

Apply that where relevant.

tbursztyka commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,602 @@
+                        }
+                }
+
+                if (data->odr > max_odr) {
```
can max_odr be below 0? 
If not, you may put this test as an else if () to previous if()

nashif commented:
@punitvara Please fix the style issues, you resolve many of the issues using uncrustify, for example:

```
$ sudo dnf install uncrustify
$ uncrustify  --no-backup -l C -c $ZEPHYR_BASE/scripts/uncrustify.cfg drivers/sensor/bmm150/bmm150.{h,c}
```

bogdan-davidoaia commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,606 @@
+        u8_t chip_id;
+        struct bmm150_preset preset;
+
+        bmm150_set_power_mode(dev, BMM150_POWER_MODE_NORMAL, 0);
```
Check return values for these too.

punitvara replied:
Done

bogdan-davidoaia commented on drivers/sensor/bmm150/bmm150.h:
```diff
@@ -0,0 +1,174 @@
+/* sensor_bmm150.h - header file for BMM150 Geomagnetic sensor driver */
+
+/*
+ * * Copyright (c) 2016 Intel Corporation
```
2017

punitvara replied:
Done

bogdan-davidoaia commented on drivers/sensor/bmm150/bmm150.h:
```diff
@@ -0,0 +1,174 @@
+#include <kernel.h>
+#include <device.h>
+#include <sensor.h>
+#include <misc/byteorder.h> /* Added for sys_le16_to_cpu */
```
Don't think you need that comment.

punitvara replied:
Done

bogdan-davidoaia commented on samples/sensor/bmm150/prj.conf:
```diff
@@ -0,0 +1,12 @@
+CONFIG_ARC_GDB_ENABLE=y
+CONFIG_BMM150_SET_ATTR=y
+CONFIG_BMM150_SAMPLING_RATE_RUNTIME=y
+#CONFIG_BMM150_TRIGGER=y
```
Remove these two commented configs.

punitvara replied:
Done

punitvara commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,606 @@
+
+static void bmm150_convert(struct sensor_value *val, int raw_val)
+{
+	/* val = raw_val /1600 */
```
Done

punitvara commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,606 @@
+		return -EIO;
+	}
+ 
+	raw_x = (s16_t)sys_le16_to_cpu(values[BMM150_AXIS_X]) >>
```
I am using raw_x as argument in bmm150_compensate_xy(). That's why raw_x cannot be directly assigned to drv_data->sample_x

punitvara commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,606 @@
+{
+	struct bmm150_data *data = dev->driver->data;
+	const struct bmm150_config *config = dev->config->config_info;
+	int reg_val;
```
done

punitvara commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,606 @@
+
+	return 0;
+}
+#endif
```
done

punitvara commented on drivers/sensor/bmm150/Kconfig:
```diff
@@ -0,0 +1,81 @@
+	default 0x10
+	depends on BMM150
+	help
+	  Specify the I2C slave address for the BMM150
```
done

punitvara committed:
driver: sensor: Add BMM150 Geomagnetic sensor driver

This patch implements BMM150 driver with polling mode and its tested
with original BMM150 sensor hardware. Driver works on I2C
interface as of now.

Following datasheet is taken as reference while developing driver.
http://www.mouser.com/ds/2/783/BST-BMM150-DS001-01-786480.pdf

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
samples: bmm150: Add sample application for BMM150 sensor

To test BMM150 geomagnetic sensor driver in polling mode,
sample application added by this patch can be used.

Signed-off-by: Punit Vara <punit.vara@intel.com>

dbkinder commented on samples/sensor/bmm150/README.rst:
```diff
@@ -0,0 +1,8 @@
+#########################
+
+Overview
+--------
```
Level 2 headings use ******
(Order is #####, *******, =======, -------- )

punitvara replied:
Document is expanded as you suggested in apds9960

dbkinder reviewed (changes requested):
Needs more documentation (see sample template)

punitvara commented on drivers/sensor/bmm150/Kconfig:
```diff
@@ -0,0 +1,96 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
Done

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,588 @@
+		((((s32_t)val) * ((s32_t)val)) >> 7)) + (((s32_t)val) *
+		((s32_t)(((s16_t)tregs->xy1) << 7)))) >> 9) +
+		((s32_t)0x100000)) * ((s32_t)(((s16_t)txy2) +
+		((s16_t)0xA0)))) >> 12)) >> 13)) + (((s16_t)txy1) << 3);
```
Please simplify these equations. There are too many casts, shifts, and magic numbers to be readable.

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,588 @@
+	val = (((((s32_t)(z - tregs->z4)) << 15) - ((((s32_t)tregs->z3) *
+		((s32_t)(((s16_t)rhall) - ((s16_t)tregs->xyz1)))) >> 2)) /
+		(tregs->z2 + ((s16_t)(((((s32_t)tregs->z1) *
+		((((s16_t)rhall) << 1))) + (1 << 15)) >> 16))));
```
Please simplify this equation

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,588 @@
+	if (i2c_burst_read(drv_data->i2c, config->i2c_slave_addr,
+			   BMM150_REG_X_L, (u8_t *)values,
+			   sizeof(values)) < 0) {
+		SYS_LOG_DBG("failed to read sample");
```
Use SYS_LOG_ERR instead of SYS_LOG_DBG. Add more of these throughout the driver when you make other I2C transactions.

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,588 @@
+	drv_data->sample_y = bmm150_compensate_xy(&drv_data->tregs,
+						raw_y, rhall, false);
+	drv_data->sample_z = bmm150_compensate_z(&drv_data->tregs,
+							raw_z, rhall);
```
Fix alignment on the second line of these three calls to bmm150_compensate

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,588 @@
+	struct bmm150_preset preset;
+
+	if (bmm150_set_power_mode(dev, BMM150_POWER_MODE_NORMAL, 0) < 0) {
+		SYS_LOG_DBG("failed to bring up device from normal mode");
```
All the SYS_LOG_DBG calls in this function should be replaced with SYS_LOG_ERR

MaureenHelm commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,588 @@
+
+DEVICE_AND_API_INIT(bmm150, CONFIG_BMM150_DEV_NAME, bmm150_init,
+		&bmm150_data, &bmm150_config, POST_KERNEL,
+		CONFIG_SENSOR_INIT_PRIORITY, &bmm150_api_funcs);
```
Fix alignment on the second two lines

dbkinder reviewed (changes requested):
Sample doc ( samples/sensor/bmm150/README.rst ) still needs to be written beyond the intro sentence

punitvara commented:
@dbkinder  Done. Can you please check ?

punitvara commented on drivers/sensor/bmm150/bmm150.c:
```diff
@@ -0,0 +1,595 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Done

dbkinder commented on samples/sensor/bmm150/README.rst:
```diff
@@ -0,0 +1,38 @@
+
+Sample application that periodically reads magnetometer (X, Y, Z) data from
+the first available device that implements SENSOR_CHAN_MAGN_* (predefined array
+of device names).This sample checks the sensor in polling mode (without
```
needs a space after the period.

punitvara replied:
Done

dbkinder commented on samples/sensor/bmm150/README.rst:
```diff
@@ -0,0 +1,38 @@
+Overview
+********
+
+Sample application that periodically reads magnetometer (X, Y, Z) data from
```
Make this a sentence.  How about starting with, 

     This sample application periodically reads ...

punitvara replied:
Done

punitvara replied:
@dbkinder What do you think about READ me ?

dbkinder reviewed (changes requested):
a couple of typos to fix please.

punitvara commented:
@nashif  Can you please review this driver ?

dbkinder commented on samples/sensor/bmm150/README.rst:
```diff
@@ -0,0 +1,38 @@
+
+.. code-block:: console
+
+        $ cd $ZEPHYR_BASE/samples/sensors/bmm150
```
samples/sensor/bmm150

dbkinder commented on samples/sensor/bmm150/README.rst:
```diff
@@ -0,0 +1,38 @@
+Building and Running
+********************
+
+This sample application uses an BMM150 sensor connected to an Arduino 101 board via I2C.
```
Can you provide a reference to how to connect the sensor, and for that matter, what other hardware is needed to flash the Arduino 101 in the steps below.  Is a USB cable sufficient?

punitvara replied:
USB and I2C is enough for connection. and for extra pins I added datasheet reference

dbkinder reviewed (changes requested):
sample instructions are rather incomplete for someone unfamiliar with Zephyr development

punitvara commented:
@nashif  I added equation as per your suggestion. Please let me know if you think something else is needed to be done.
